### DISCUSSION TITLE
Problems with modifiers(&SecurityAddon)

### DISCUSSION THREAD

**Discussion by nelsonsilva94** [juhaku/utoipa#421] [General]
Hi,

adding the `modifiers(&SecurityAddon)` seems to get schemas errors that previously were correct.

Before `modifiers(&SecurityAddon)`:

![image](https://user-images.githubusercontent.com/63600033/209683037-b4a74f93-47a9-4b38-81b7-35929546d027.png)

After adding `modifiers(&SecurityAddon)` line
```rust
    #[derive(OpenApi)]
    #[openapi(
        paths(
            auth::hello,
            auth::login,
            auth::auth_refresh
        ),
        components(
            schemas(auth::UserRecord, auth::AuthError, auth::User, auth::UserCredentials)
        ),
        tags(
            (name = "pocketbase-auth", description = "Auth with pocketbase.")
        ),
        modifiers(&SecurityAddon)
    )]
```

I got the following errors:

![image](https://user-images.githubusercontent.com/63600033/209683601-76482746-bad4-4b14-adfc-6fcef766afdb.png)

All the schemas (auth::UserCredentials, auth::User, auth::AuthError) were Okay before.

Am I missing something pretty obvious?

Thanks in advance,
Cheers

**Comment by juhaku** [comment]
What happens in your modifier?
```rust
struct MyModifier;

impl Modify for MyModifier {
  fn modify(&self, &mut openapi: OpenApi) {
    // Get access to the existing components or insert new one if none is present. 
    // If there already is components added with #[openapi(schemas(...))] then the following line should
    // return you the existing components list that you can modify.
    // Do not override it with new components otherwise you will loose everything that has been already added there.
    let components = openapi.components.get_or_insert(Components::new());
  }
}
```

Alternatively you can directly use the `OpenApi` type to modify the generated doc.
```rust
#[derive(OpenApi)]
struct ApiDoc;

let mut openapi = ApiDoc::openapi();
let mut componets = openapi.componets.get_or_insert(Components::new());
```

**Reply by nelsonsilva94** [reply]
My modifier looks like this: 
```rust
    struct SecurityAddon;

    impl Modify for SecurityAddon {
        fn modify(&self, openapi: &mut utoipa::openapi::OpenApi) {
            openapi.components = Some(
                utoipa::openapi::ComponentsBuilder::new()
                    .security_scheme(
                        "api_auth_token",
                        SecurityScheme::Http(
                            HttpBuilder::new()
                                .scheme(HttpAuthScheme::Bearer)
                                .bearer_format("JWT")
                                .build(),
                        ),
                    )
                    .build(),
            )
        }
    }
```

**Reply by juhaku** [reply]
> ```rust
> openapi.components = Some(...)
> ```
This is the issue, that overrides all the components. The modifier will modify the existing openapi that is generated when macros compile.

**Reply by nelsonsilva94** [reply]
I see, then how can I make it work?

Using 
```rust
openapi.componets.get_or_insert(...)
```
 instead of 
```rust
openapi.components = Some(...)
```
Is this the way?

**Reply by juhaku** [reply]
If you have already components registered via `[openapi(...)]` macro attribute as seen below
>```rust
>    #[derive(OpenApi)]
>    #[openapi(
>        // ....
>        components(
>            schemas(auth::UserRecord, auth::AuthError, auth::User, auth::UserCredentials)
>        ),
>        // ....
>        modifiers(&SecurityAddon)
>    )]
>```

The `openapi.components` already contains components. When you call `openapi.components = Some()..` you are assigning it a new value instead of mutating the existing one. If you add call `dbg!(&openapi.components);` within your `fn modify() {}` function body before you assign a new value, you see that there already is AuthError, UserRecord, etc.

The modifier will not create a new OpenApi instance and then merge the values from it to a single one which contains all the components, but instead it will modify the same instance you use to add items via declaring them via macro attributes. That is also why the `openapi` argument is `openapi: &mut OpenApi` it is mutable variable.

**Reply by nelsonsilva94** [reply]
Thanks for the help!
I was able to make it work with:
```rust
  impl Modify for SecurityAddon {
        fn modify(&self, openapi: &mut utoipa::openapi::OpenApi) {
            let components = openapi.components.as_mut().unwrap(); // we can unwrap safely since there already is components registered.
            components.add_security_scheme(
                "api_auth_token",
                SecurityScheme::Http(
                    HttpBuilder::new()
                        .scheme(HttpAuthScheme::Bearer)
                        .bearer_format("JWT")
                        .build(),
                ),
            )
        }
    }
```

Is this the correct approach?

**Reply by juhaku** [reply]
Yes, that will work as well because you already have components in OpenApi registered. If you hadn't had defined schemas within `openapi(...)`  attribute the unwrap would then fail in panic.
